### PR TITLE
Fixed data race in all_type_info in free-threading mode

### PR DESCRIPTION
**Pull request by vfdev-5**
## Description
- fixed data race in `all_type_info` in free-threading mode
- added test

For example, we have 2 threads entering `all_type_info`. Both enter `all_type_info_get_cache`` function and there is a first one which inserts a tuple (type, empty_vector) to the map and second is waiting. Inserting thread gets the (iter_to_key, True) and non-inserting thread after waiting gets (iter_to_key, False).
Inserting thread than will add a weakref and will then call into `all_type_info_populate`. However, non-inserting thread is not entering `if (ins.second) {` clause and returns `ins.first->second;`` which is just empty_vector. Finally, non-inserting thread is failing the check in `allocate_layout`:
```c++
if (n_types == 0) {
    pybind11_fail(
        "instance allocation failed: new instance has no pybind11-registered base types");
}
```

On `master` running this test gives:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  instance allocation failed: new instance has no pybind11-registered base types
Fatal Python error: Aborted
```

## Suggested changelog entry:

```rst
A free-threading data race in ``all_type_info()`` was fixed.
```

cc @colesbury

### CODE REVIEW

**Comment by colesbury** [comment]
I think we may need a further rework of `all_type_info()`. It's used by `get_type_info()` and some of the callers to that function modify the returned `detail::type_info*`.

I'm not entirely sure what the locking strategy should be here. It's possible that all the callers to `get_type_info()` and `all_type_info()` should hold the internals lock.

**Comment by vfdev-5** [comment]
@rwgk can you please review this PR?

**Comment by rwgk** [comment]
Will try tomorrow.

@cryos could you please also take a look?

File: include/pybind11/detail/type_caster_base.h
Comment (rwgk) [inline include/pybind11/detail/type_caster_base.h:178]: (Minor) I'd go all the way:

```
    return all_type_info_get_cache(type).first->second;
```

File: include/pybind11/pybind11.h
Comment (rwgk) [inline include/pybind11/pybind11.h:2339]: Suggested rewording (IIUC):

// For free-threading mode, this call must be under the with_internals() mutex lock, to avoid that other threads continue running with the empty ins.first->second.

File: tests/pybind11_tests.h
Comment (rwgk) [inline tests/pybind11_tests.h:100]: Could you please add a comment:

// See PR #5419 for background.

(That'll make it much faster for people to find out what this code is about.)

File: tests/test_class.py
Comment (rwgk) [inline tests/test_class.py:529]: Suggestion to replace this very long comment with simply:
```
# See PR #5419 for background.
```
This suggestion is mainly because the explanation applies to the code **before** this PR. I think that'll be confusing. It's better to send people to the PR description.

File: tests/test_class.py
Comment (rwgk) [inline tests/test_class.py:536]: Could this just be

```
        pass
```

?

If not, could you please add a terse comment to state that the explicit `__init__` is critical? (I think it's fine to omit explaining why, if it's complicated to explain. The comment is to reassure people that this is really needed.)

**Review by rwgk — COMMENTED** [review]
@colesbury could you please formally approve?

Looks good to me, but I'm not a free-threading expert. I'll merge when I see approvals from @colesbury and @cryos.

File: tests/test_class.py
Comment (vfdev-5) [reply tests/test_class.py:536]: I do not why exactly but this code:
```python
class Context(TestContext):
    pass
```
in the test does not make `master` to fail. But 
```python
    class Context(TestContext):
        def __init__(self, *args, **kwargs):
            super().__init__(*args, **kwargs)
```
makes it fail.

**Review by rwgk — APPROVED** [review]
Looks good to me, but waiting for approvals from @colesbury and @cryos.

**Review by cryos — APPROVED** [review]
This looks good to me.

**Pull request merged**
